### PR TITLE
New `bytes_from_field_elements` util

### PR DESCRIPTION
**Pull request by ggutoski**
## Description

closes: #227 

- add `bytes_from_field_elements()` to invert `bytes_to_field_elements()`
- modify `bytes_to_field_elements()` to encode bytes length in the first field element for use in inversion
- remove unneeded bytes copy in `bytes_to_field_elements` and other tidying
- generalize both functions from `PrimeField` to `Field`
- TODO: used `String` as error type for `bytes_from_field_elements`---should we do something else instead?
- minimal tests only

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Review by mrain — DISMISSED** [review]
LGTM

**Review by chancharles92 — COMMENTED** [review]
LGTM, a few minor comments. There is also a minor clippy warning.

File: utilities/src/conversion.rs [hunk 1]
```diff
@@ -93,35 +95,121 @@ where
...
+    //   sufficiently small
+    // - collect PrimeField elements into Field elements and append to result
+    let primefield_chunk_len = ((F::BasePrimeField::MODULUS_BIT_SIZE - 1) / 8) as usize;
+    let extension_degree = F::extension_degree() as usize;
+    let field_chunk_len = primefield_chunk_len * extension_degree;
+    let result_length = (bytes.as_ref().len() + field_chunk_len - 1) / field_chunk_len + 1;
+    let mut result = Vec::with_capacity(result_length);
+
+    // the first field element encodes the bytes length as u64
+    result.push(F::from(bytes.as_ref().len() as u64));
```
[the remark this answers is not in the document]
Comment (ggutoski) [reply utilities/src/conversion.rs:118]: Oh right I was going to add a check for that but forgot. Or maybe we can use trait bounds to enforce sufficiently large field characteristic? In the worst case we could just panic if characteristic is too small. I don't think we should spend too much more time on this.

File: utilities/src/conversion.rs [hunk 2]
```diff
@@ -93,35 +95,121 @@ where
...
-    let mut result = Vec::new();
-    for chunk in padded_bytes.chunks(chunk_length) {
-        result.push(F::from_le_bytes_mod_order(chunk));
+    // - partition bytes into chunks of length one fewer than the base prime field
+    //   modulus byte length
+    // - convert each chunk into PrimeField via from_le_bytes_mod_order
+    // - modular reduction is guaranteed not to occur because chunk byte length is
+    //   sufficiently small
+    // - collect PrimeField elements into Field elements and append to result
+    let primefield_chunk_len = ((F::BasePrimeField::MODULUS_BIT_SIZE - 1) / 8) as usize;
```
[the remark this answers is not in the document]
Comment (mrain) [reply utilities/src/conversion.rs:111]: I'm not sure. But we could add a special case when `BasePrimeField` is GF_2.

**Comment by ggutoski** [comment]
> LGTM, a few minor comments. There is also a minor clippy warning.

Found the clippy warning and fixed it. It was in `plonk`, so not my fault. 😉

**Review by chancharles92 — APPROVED** [review]

File: utilities/src/conversion.rs [hunk 3]
```diff
@@ -93,35 +95,133 @@ where
...
+        .get(..mem::size_of::<u64>())
+        .ok_or("can't read result len from field element: not enough bytes")?;
+    let result_len = u64::from_le_bytes(first_elem_bytes.try_into().unwrap());
+    let result_len =
+        usize::try_from(result_len).map_err(|_| "can't convert result len u64 to usize")?;
+
+    let primefield_chunk_len = ((F::BasePrimeField::MODULUS_BIT_SIZE - 1) / 8) as usize;
+    let extension_degree = F::extension_degree() as usize;
+    let field_chunk_len = primefield_chunk_len * extension_degree;
+    let result_capacity = elems.len() * field_chunk_len;
```
[the remark this answers is not in the document]
Comment (ggutoski) [reply utilities/src/conversion.rs:179]: done in https://github.com/EspressoSystems/jellyfish/pull/255/commits/7e46a86e4cd1fa32e4c623ade6d2b5c596ac1c72

File: utilities/src/conversion.rs [hunk 4]
```diff
@@ -161,4 +261,30 @@ mod tests {
...
+        let elems: Vec<F> = bytes_to_field_elements(&bytes);
+        let result = bytes_from_field_elements(&elems).unwrap();
+        assert_eq!(result, bytes);
+    }
+
+    #[test]
+    fn test_bytes_field_elems() {
+        bytes_field_elems::<Fr381>();
+        bytes_field_elems::<Fr254>();
+    }
```
[the remark this answers is not in the document]
Comment (ggutoski) [reply utilities/src/conversion.rs:289]: done in https://github.com/EspressoSystems/jellyfish/pull/255/commits/f1327a17f2a709beaaa6c081948766d97fa37462

**Pull request merged**